kaldb/src/main/java/com/slack/kaldb/logstore/search/KaldbDistributedQueryService.java

```diff
@@ -126,6 +136,135 @@ private KaldbServiceGrpc.KaldbServiceFutureStub getKaldbServiceGrpcClient(String
+
+  public static boolean isSnapshotInPartition(
+      SnapshotMetadata snapshotMetadata,
+      ServiceMetadataStore serviceMetadataStore,
+      long queryStartTimeEpochMs,
+      long queryEndTimeEpochMs,
+      String indexName) {
+    List<ServicePartitionMetadata> partitions =
+        findPartitionsToQuery(
+            serviceMetadataStore, queryStartTimeEpochMs, queryEndTimeEpochMs, indexName);
+    return partitions
+        .stream()
+        .anyMatch(
+            partitionMetadata ->
+                partitionMetadata.partitions.contains(snapshotMetadata.partitionId)
+                    && containsDataInTimeRange(
+                        partitionMetadata.startTimeEpochMs,
+                        partitionMetadata.endTimeEpochMs,
+                        snapshotMetadata.startTimeEpochMs,
+                        snapshotMetadata.endTimeEpochMs));
+  }
+
+  private static String getSnapshotName(SearchMetadata searchMetadata) {
+    return searchMetadata.snapshotName.startsWith("LIVE")
+        ? searchMetadata.snapshotName.substring(5) // LIVE_
+        : searchMetadata.snapshotName;
+  }
+
+  /*
+   If there is only one node hosting the snapshot use that
+   If there are multiple, then always prefer cache nodes. We pick any cache node at random
+  */
+  public static String pickSearchNodeToQuery(List<SearchMetadata> queryableSearchMetadataNodes) {
+    if (queryableSearchMetadataNodes.size() == 1) {
+      return queryableSearchMetadataNodes.get(0).url;
+    } else {
+      List<SearchMetadata> cacheNodeHostedSearchMetadata =
+          queryableSearchMetadataNodes
+              .stream()
+              .filter(searchMetadata -> !searchMetadata.isLive)
+              .collect(Collectors.toList());
+      if (cacheNodeHostedSearchMetadata.size() == 1) {
+        return cacheNodeHostedSearchMetadata.get(0).url;
+      } else {
+        return cacheNodeHostedSearchMetadata.get(
+                ThreadLocalRandom.current().nextInt(cacheNodeHostedSearchMetadata.size()))
+            .url;
+      }
+    }
+  }
+
+  private KaldbServiceGrpc.KaldbServiceFutureStub getStub(String url) {
+    if (stubs.get(url) != null) {
+      return stubs.get(url);
+    } else {
+      LOG.warn(
+          "snapshot {} is not cached. ZK listener on searchMetadataStore should have cached the stub",
+          url);
+      return null;
+    }
+  }
+
+  /*
+   get partitions that match on two criteria
+   1. index name
+   2. partitions that have an overlap with the query window
+  */
+  @VisibleForTesting
+  public static List<ServicePartitionMetadata> findPartitionsToQuery(
+      ServiceMetadataStore serviceMetadataStore,
+      long startTimeEpochMs,
+      long endTimeEpochMs,
+      String indexName) {
+    return serviceMetadataStore
+        .getCached()
+        .stream()
+        .filter(serviceMetadata -> serviceMetadata.name.equals(indexName))
+        .flatMap(serviceMetadata -> serviceMetadata.partitionConfigs.stream())
+        .filter(
```

Comment:
I think the flatMap is deliberate to map serviceMetadata to it's partition configs. 

Do you wanna try it and see if it works? I'm not sure how it would work if I swap it like you suggested.